File: frontend/src/types/evaluation.ts
export type EvaluationMode = 'retrieval_only' | 'rag_full';
export type EvaluationStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
export type ScoringMethod = 'semantic_similarity' | 'llm_judge';
export type FailureType =
  | 'mcp_error'
  | 'no_recall'
  | 'wrong_answer'
  | 'hallucination'
  | 'incomplete'
  | 'mapping_error';

export interface EvaluationConfig {
  top_k: number;
  pass_threshold: number;
  scoring_method: ScoringMethod;
  mcp_timeout_ms: number;
}

export interface Evaluation {
  id: string;
  name: string;
  kb_id: string;
  kb_name: string;
  question_bank_id: string;
  question_bank_name: string;
  mode: EvaluationMode;
  config: EvaluationConfig;
  status: EvaluationStatus;
  progress: number;
  completed_questions: number;
  total_questions: number;
  overall_score: number | null;
  pass_rate: number | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface EvaluationListResponse {
  items: Evaluation[];
  total: number;
}

export interface EvaluationListFilters {
  status?: EvaluationStatus;
  search?: string;
}

export interface EvaluationCreatePayload {
  name: string;
  kb_id: string;
  question_bank_id: string;
  mode?: EvaluationMode;
  config?: Partial<EvaluationConfig>;
}

export interface ChunkResult {
  content: string;
  source: string;
  title?: string | null;
  score?: number | null;
}

export interface EvaluationResult {
  id: string;
  question_id: string;
  question: string;
  expected_answer: string;
  category?: string | null;
  difficulty?: string | null;
  mcp_tool: string;
  mcp_request: Record<string, unknown>;
  mcp_response_raw?: Record<string, unknown> | null;
  chunks: ChunkResult[];
  generated_answer?: string | null;
  score: number;
  passed: boolean;
  failure_type?: FailureType | null;
  failure_reason?: string | null;
  latency_ms: number;
  created_at: string;
}

export interface EvaluationResultsResponse {
  items: EvaluationResult[];
  total: number;
  page: number;
  page_size: number;
}

export interface CategoryScore {
  category: string;
  score: number;
  pass_rate: number;
  total: number;
}

export interface ReportSummary {
  overall_score: number;
  pass_rate: number;
  recall_rate: number;
  mcp_success_rate: number;
  avg_mcp_latency_ms: number;
  total_questions: number;
  passed: number;
  failed: number;
}

export interface EvaluationReport {
  task_id: string;
  task_name: string;
  kb_name: string;
  question_bank_name: string;
  mode: EvaluationMode;
  status: EvaluationStatus;
  completed_at: string | null;
  summary: ReportSummary;
  failure_distribution: Record<string, number>;
  category_scores: CategoryScore[];
}

export const FAILURE_TYPE_LABELS: Record<FailureType, string> = {
  mcp_error: 'MCP 错误',
  no_recall: '未召回',
  wrong_answer: '答案错误',
  hallucination: '幻觉',
  incomplete: '不完整',
  mapping_error: '映射错误',
};

export const EVALUATION_STATUS_LABELS: Record<EvaluationStatus, string> = {
  pending: '等待中',
  running: '运行中',
  completed: '已完成',
  failed: '失败',
  cancelled: '已取消',
};

File: backend/app/schemas/evaluation.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

EvaluationMode = Literal["retrieval_only", "rag_full"]
EvaluationStatus = Literal["pending", "running", "completed", "failed", "cancelled"]
ScoringMethod = Literal["semantic_similarity", "llm_judge"]
FailureType = Literal[
    "mcp_error",
    "no_recall",
    "wrong_answer",
    "hallucination",
    "incomplete",
    "mapping_error",
]


class EvaluationConfig(BaseModel):
    top_k: int = Field(default=5, ge=1, le=50)
    pass_threshold: float = Field(default=70, ge=0, le=100)
    scoring_method: ScoringMethod = "semantic_similarity"
    mcp_timeout_ms: int = Field(default=30000, ge=1000, le=120000)


class EvaluationCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    kb_id: str
    question_bank_id: str
    mode: EvaluationMode = "retrieval_only"
    config: EvaluationConfig = Field(default_factory=EvaluationConfig)


class EvaluationResponse(BaseModel):
    id: str
    name: str
    kb_id: str
    kb_name: str
    question_bank_id: str
    question_bank_name: str
    mode: EvaluationMode
    config: EvaluationConfig
    status: EvaluationStatus
    progress: int
    completed_questions: int
    total_questions: int
    overall_score: float | None = None
    pass_rate: float | None = None
    error_message: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None


class EvaluationListResponse(BaseModel):
    items: list[EvaluationResponse]
    total: int


class ChunkResult(BaseModel):
    content: str
    source: str
    title: str | None = None
    score: float | None = None


class EvaluationResultResponse(BaseModel):
    id: str
    question_id: str
    question: str
    expected_answer: str
    category: str | None = None
    difficulty: str | None = None
    mcp_tool: str
    mcp_request: dict
    mcp_response_raw: dict | None = None
    chunks: list[ChunkResult]
    generated_answer: str | None = None
    score: float
    passed: bool
    failure_type: FailureType | None = None
    failure_reason: str | None = None
    latency_ms: int
    created_at: datetime


class EvaluationResultsListResponse(BaseModel):
    items: list[EvaluationResultResponse]
    total: int
    page: int
    page_size: int


class CategoryScore(BaseModel):
    category: str
    score: float
    pass_rate: float
    total: int


class ReportSummary(BaseModel):
    overall_score: float
    pass_rate: float
    recall_rate: float
    mcp_success_rate: float
    avg_mcp_latency_ms: float
    total_questions: int
    passed: int
    failed: int


class EvaluationReportResponse(BaseModel):
    task_id: str
    task_name: str
    kb_name: str
    question_bank_name: str
    mode: EvaluationMode
    status: EvaluationStatus
    completed_at: datetime | None
    summary: ReportSummary
    failure_distribution: dict[str, int]
    category_scores: list[CategoryScore]
